src/App.tsx

```diff
@@ -1,223 +1,154 @@
+    setErrorMessages((prev) => [message, ...prev]);
+
+    setTimeout(() => closeErrorMessage(message), 3000);
+  }, [closeErrorMessage]);
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    if (user) {
+      todoApi.getTodos(user.id)
+        .then(setTodos)
+        .catch(() => showError('Todos loading failed'));
     }
-  }, []);
+  }, [user]);
+
+  const onAddTodo = useCallback(async (fieldsForCreate: Omit<Todo, 'id'>) => {
+    try {
+      setIsAddingTodo(true);
+      setTempTodo({
+        ...fieldsForCreate,
+        id: 0,
+      });
+
+      const newTodo = await todoApi.createTodo(fieldsForCreate);
+
+      setTodos(prev => [...prev, newTodo]);
+    } catch {
+      showError('Unable to add a todo');
+
+      throw Error('Error while adding todo');
+    } finally {
+      setTempTodo(null);
+      setIsAddingTodo(false);
+    }
+  }, [showError]);
+
+  const onDeleteTodo = useCallback(async (todoId: number) => {
+    try {
+      setDeletingTodoIds(prev => [...prev, todoId]);
+
+      await todoApi.deleteTodo(todoId);
+
+      setTodos(prev => prev.filter(todo => todo.id !== todoId));
+    } catch {
+      showError('Unable to delete todo');
+    } finally {
+      setDeletingTodoIds(prev => prev.filter(id => id !== todoId));
+    }
+  }, [showError]);
+
+  const onDeleteCompleted = useCallback(async () => {
+    const completedTodoIds = todos.filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    completedTodoIds.forEach(id => onDeleteTodo(id));
+  }, [onDeleteTodo, todos]);
+
+  const activeTodos = useMemo(() => (
+    todos.filter(todo => !todo.completed)
+  ), [todos, completedFilter]);
+
+  const visibleTodos = useMemo(() => {
+    if (completedFilter === CompletedFilter.All) {
+      return todos;
+    }
+
+    return todos.filter((todo) => (CompletedFilter.Completed === completedFilter
+      ? todo.completed
+      : !todo.completed
+    ));
+  }, [todos, completedFilter]);
+
+  const shouldRenderContent = todos.length > 0 || !!tempTodo;
 
   return (
     <div className="todoapp">
       <h1 className="todoapp__title">todos</h1>
 
       <div className="todoapp__content">
-        <header className="todoapp__header">
-          <button
-            data-cy="ToggleAllButton"
-            type="button"
-            className="todoapp__toggle-all active"
-          />
-
-          <form>
-            <input
-              data-cy="NewTodoField"
-              type="text"
-              ref={newTodoField}
-              className="todoapp__new-todo"
-              placeholder="What needs to be done?"
+        <Header
+          onAddTodo={onAddTodo}
+          isAddingTodo={isAddingTodo}
+          showError={showError}
+        />
+
+        {shouldRenderContent && (
+          <>
+            <TodoList
+              todos={visibleTodos}
+              tempTodo={tempTodo}
+              onDeleteTodo={onDeleteTodo}
```

Comment:
Your function name should sound like a call to action, to do something
For example `onClick` will `deleteTodo`
Fix it here in all other places

```suggestion
              deleteTodo={deleteTodo}
```